### PR TITLE
[13.0][ADD] pricelist_items_auto_mailing

### PR DESCRIPTION
**Pull request by ChristianSantamaria**

### CODE REVIEW

**Comment by Anxo82** [comment]
Please review.
Thanks.

**Review by dalonsod — CHANGES_REQUESTED** [review]
@ChristianSantamaria @Anxo82 we could have an in-depth review together and look at these potential problems/things that could needed to be fixed AFAIK:
- Main addon `product_pricelist_auto_maling` could be focused in pricelist items changes rather than pricelist header, so `product_pricelist_mgmt_advanced` dependency should become unneeded, because `product_pricelist_mgmt_advanced_auto_mailing` is no longer needed (not sure).
- If `product_pricelist_mgmt_advanced_auto_mailing` is still needed, it should be moved to `solvosci/slv-product-priv` repo and `oca_dependencies.txt` addition removed anyway.
- Current behavior implies a mail sending during pricelist update. This could lead to a poor performance, because every pricelist update will wait until a mail is sent. Such mail process should be moved to a cron IMO. Then, the problem becomes determining which pricelists (more precise, their items) have changed since last cron execution, and determining it in an efficient way. We'll talk about it.

**Comment by ChristianSantamaria** [comment]
@dalonsod can u review?

File: oca_dependencies.txt [hunk 1]
```diff
@@ -0,0 +1,2 @@
+# See https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst#oca_dependencies-txt
+slv-product-priv https://github.com/solvosci/slv-product-priv
```
Comment (dalonsod) [inline oca_dependencies.txt]: Could this dependency be removed?

File: product_pricelist_auto_mailing/models/product_pricelist.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+        mailable_pricelist_ids = self.env['product.pricelist'].search([('mailable_user_ids', '!=', False)])
+        if mailable_pricelist_ids:
+            body = []
+            for pricelist in mailable_pricelist_ids:
+                body.append(_('The pricelist %s, has changed for users: %s') % (pricelist.name, ' ,'.join(pricelist.mailable_user_ids.mapped('name'))))
+                pricelist.mailable_user_ids = False
+            ctx = self.env.context.copy()
+            ctx.update({'body': body})
+            pricelist_template_id = self.env.ref('product_pricelist_auto_mailing.mail_template_data_pricelist_notification')
+            return pricelist_template_id.with_context(ctx).send_mail(mailable_pricelist_ids[0].id, force_send=True)
```
Comment (dalonsod) [inline product_pricelist_auto_mailing/models/product_pricelist.py:23]: I figure that only one of the pricelists will reflect in its thread a mail send for every cron execution, is that correct?

File: product_pricelist_auto_mailing/models/product_pricelist.py [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+# License LGPL-3 - See http://www.gnu.org/licenses/lgpl-3.0.html
+
+from odoo import models, _, fields
+
+
+class ProductPricelist(models.Model):
+    _name = 'product.pricelist'
+    _inherit = ['product.pricelist', 'mail.thread', 'mail.activity.mixin']
+
+    mailable_user_ids = fields.Many2many('res.users')
```
Comment (dalonsod) [inline product_pricelist_auto_mailing/models/product_pricelist.py]: This field is a technical one, so it should be protected, and it shouldn't be copied:
```suggestion
    mailable_user_ids = fields.Many2many('res.users', readonly=True, copy=False)
```

File: product_pricelist_auto_mailing/models/product_pricelist_item.py [hunk 4]
```diff
@@ -0,0 +1,13 @@
+# © 2024 Solvos Consultoría Informática (<http://www.solvos.es>)
+# License LGPL-3 - See http://www.gnu.org/licenses/lgpl-3.0.html
+
+from odoo import models
+
+
+class ProductPricelistItem(models.Model):
+    _inherit = 'product.pricelist.item'
+
+    def write(self, vals_list):
```
Comment (dalonsod) [inline product_pricelist_auto_mailing/models/product_pricelist_item.py]: Don't use `vals_list`, as suggests a list, but this parameter is a dictionary (`vals_list` is reserved for a multi `create` method):
```suggestion
    def write(self, vals):
```

File: product_pricelist_auto_mailing/models/product_pricelist_item.py [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+        if self.env.user.id in self.env.company.user_notifying_price_changes_ids.ids:
+            self.pricelist_id.mailable_user_ids = [(4, self.env.user.id)]
```
Comment (dalonsod) [inline product_pricelist_auto_mailing/models/product_pricelist_item.py]: Not sure if this notation will always work. `write` is a multi method, so seems to be more accurate this, as `pricelist_id` could be one ore more pricelists (not tested), could you check?
```suggestion
            self.pricelist_id.write({"mailable_user_ids": [(4, self.env.user.id)]})
```

File: product_pricelist_auto_mailing/models/product_pricelist.py (hunk 3 above)
Comment (dalonsod) [inline product_pricelist_auto_mailing/models/product_pricelist.py]: Is it possible to add this field to pricelist form view, only for debug users?

**Review by dalonsod — CHANGES_REQUESTED** [review]
See comments

**Comment by dalonsod** [comment]
And a last question, could you provide a snapshot of a mail message example?

**Comment by ChristianSantamaria** [comment]
> And a last question, could you provide a snapshot of a mail message example?

imagen

File: product_pricelist_auto_mailing/models/product_pricelist.py (hunk 2 above)
Comment (ChristianSantamaria) [reply product_pricelist_auto_mailing/models/product_pricelist.py:23]: Yes

Comment (dalonsod) [reply product_pricelist_auto_mailing/models/product_pricelist.py:23]: OK, we'll simply explain it to users when needed. _It's not a bug, but a feature!_

**Comment by dalonsod** [comment]
Removed empty `oca_dependencies.txt`.

This is a beta version. Merging...

**Pull request merged**